src/main/host/descriptor/socket/abstract_unix_ns.rs

```diff
@@ -0,0 +1,273 @@
+        socket: &Arc<AtomicRefCell<UnixSocketFile>>,
+        mut rng: impl rand::Rng,
+    ) -> Result<(), BindError> {
+        let mut ns = ns_arc.borrow_mut();
+
+        // the unused name that we will bind the socket to
+        let mut name = None;
+
+        // try 10 random names
+        for _ in 0..10 {
+            let random_name: [u8; NAME_LEN] = random_name(&mut rng);
+
+            if !ns.address_map.contains_key(&random_name[..]) {
+                name = Some(random_name.to_vec());
+                break;
+            }
+        }
+
+        // if unsuccessful, try a linear search through all valid names
+        if name.is_none() {
+            for x in 0..CHARSET.len().pow(NAME_LEN as u32) {
+                let temp_name: [u8; NAME_LEN] = incremental_name(x);
+
+                if !ns.address_map.contains_key(&temp_name[..]) {
+                    name = Some(temp_name.to_vec());
+                    break;
+                }
+            }
+        }
+
+        let name = match name {
+            Some(x) => x,
+            // every valid name has been taken
+            None => return Err(BindError::NoNamesAvailable),
+        };
+
+        let name_copy = name.clone();
+
+        // when the socket closes, remove this entry from the namespace
+        let handle = Self::on_socket_close(
+            Arc::downgrade(&ns_arc),
+            &mut socket.borrow_mut(),
+            move |ns| {
+                assert!(ns.unbind(&name_copy).is_ok());
+            },
+        );
+
+        if let std::collections::hash_map::Entry::Vacant(entry) = ns.address_map.entry(name) {
+            entry.insert((Arc::downgrade(socket), handle));
+        } else {
+            unreachable!();
+        }
+
+        Ok(())
+    }
+
+    pub fn unbind(&mut self, name: &Vec<u8>) -> Result<(), BindError> {
+        // remove the namespace entry which includes the handle, so the event listener will
+        // automatically be removed from the socket
+        if self.address_map.remove(name).is_none() {
+            // didn't exist in the address map
+            return Err(BindError::NameNotFound);
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum BindError {
+    /// The name is already in use.
+    NameInUse,
+    /// Names in the ephemeral name range are all in use.
+    NoNamesAvailable,
+    /// The name was not found in the address map.
+    NameNotFound,
+}
+
+impl std::error::Error for BindError {}
+
+impl std::fmt::Display for BindError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::NameInUse => write!(f, "Name is already in use"),
+            Self::NoNamesAvailable => {
+                write!(f, "Names in the ephemeral name range are all in use")
+            }
+            Self::NameNotFound => write!(f, "Name was not found in the address map"),
+        }
+    }
+}
+
+/// The characters that are valid in auto-generated names.
```

Comment:
A comment on how these constants are chosen would be helpful, particularly if there's a man page or something to cite.

Reply:
Added.